File: frontend/src/main/resources/static/js/announcement.js
/*
 * Profit Basetool - squadron-management web app.
 * Copyright (C) 2026 Lucas Greuloch
 *
 * SPDX-License-Identifier: GPL-3.0-only
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/* global ANNOUNCE_MSG, ANNOUNCE_CONFLICT */

document.addEventListener('DOMContentLoaded', function () {
    const deleteModal = document.getElementById('delete-confirm-modal');
    const triggerDeleteBtn = document.getElementById('trigger-delete-confirm');
    const closeDeleteBtns = document.querySelectorAll('.close-delete-modal');
    const updateForm = document.getElementById('announcement-update-form');
    const deleteForm = document.getElementById('announcement-delete-form');

    function closeDeleteModal() {
        if (deleteModal) window.krtModal.close(deleteModal);
    }

    if (triggerDeleteBtn && deleteModal) {
        triggerDeleteBtn.onclick = function () {
            window.krtModal.open(deleteModal);
        };
    }

    closeDeleteBtns.forEach(function (btn) {
        btn.onclick = closeDeleteModal;
    });

    window.onclick = function (event) {
        if (event.target === deleteModal) {
            closeDeleteModal();
        }
    };

    if (updateForm) {
        updateForm.addEventListener('submit', function (event) {
            event.preventDefault();
            if (!window.krtFetch) {
                updateForm.submit();
                return;
            }
            const versionInput = updateForm.querySelector('input[name="version"]');
            const contentInput = updateForm.querySelector('textarea[name="content"]');
            window.krtFetch.write({
                method: 'POST',
                url: updateForm.getAttribute('action'),
                payload: {
                    content: contentInput ? contentInput.value : '',
                    version:
                        versionInput && versionInput.value !== ''
                            ? Number(versionInput.value)
                            : null,
                },
                successMessage: ANNOUNCE_MSG.saveSuccess,
                errorMessage: ANNOUNCE_MSG.saveError,
                conflict: ANNOUNCE_CONFLICT,
                onSuccess(body) {
                    if (versionInput && body && body.version != null) {
                        versionInput.value = body.version;
                    }
                },
            });
        });
    }

    if (deleteForm) {
        deleteForm.addEventListener('submit', function (event) {
            event.preventDefault();
            if (!window.krtFetch) {
                deleteForm.submit();
                return;
            }
            window.krtFetch.write({
                method: 'POST',
                url: deleteForm.getAttribute('action'),
                successMessage: ANNOUNCE_MSG.deleteSuccess,
                errorMessage: ANNOUNCE_MSG.deleteError,
                conflict: ANNOUNCE_CONFLICT,
                onSuccess() {
                    if (updateForm) {
                        const versionInput = updateForm.querySelector('input[name="version"]');
                        const contentInput = updateForm.querySelector('textarea[name="content"]');
                        if (contentInput) contentInput.value = '';
                        if (versionInput) versionInput.value = '0';
                    }
                    closeDeleteModal();
                },
            });
        });
    }
});

File: backend/src/main/java/de/greluc/krt/profit/basetool/backend/model/Announcement.java
/*
 * Profit Basetool - squadron-management web app.
 * Copyright (C) 2026 Lucas Greuloch
 *
 * SPDX-License-Identifier: GPL-3.0-only
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package de.greluc.krt.profit.basetool.backend.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * Announcement JPA entity.
 *
 * <p>The {@code updatedAt} column lives on {@link AbstractEntity} where {@code @UpdateTimestamp}
 * keeps it fresh on every {@code persist}/{@code update}. Earlier revisions shadowed that field
 * here with a manual {@code @PrePersist}/{@code @PreUpdate} hook — the duplicate has been removed
 * to avoid the JPA column-mapping ambiguity and to silence the CodeQL "missing {@code @Override} on
 * {@code getUpdatedAt}" finding that the shadowing produced.
 */
@Entity
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Announcement extends AbstractEntity<UUID> {

  @Getter(onMethod_ = @__(@Override))
  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  private UUID id;

  @Getter
  @Column(columnDefinition = "TEXT")
  private String content;
}
